beacon_node/beacon_chain/src/sync_committee_verification.rs

```diff
@@ -0,0 +1,726 @@
+            .map_err(BeaconChainError::from)?
+        {
+            return Err(Error::PriorSyncSignatureKnown {
+                validator_index: aggregator_index,
+                slot: contribution.slot,
+            });
+        }
+        Ok(VerifiedSyncContribution { signed_aggregate })
+    }
+
+    /// A helper function to add this aggregate to `beacon_chain.op_pool`.
+    pub fn add_to_pool(self, chain: &BeaconChain<T>) -> Result<(), Error> {
+        chain.add_contribution_to_block_inclusion_pool(self)
+    }
+
+    /// Returns the underlying `contribution` for the `signed_aggregate`.
+    pub fn contribution(self) -> SyncCommitteeContribution<T::EthSpec> {
+        self.signed_aggregate.message.contribution
+    }
+
+    /// Returns the underlying `signed_aggregate`.
+    pub fn aggregate(&self) -> &SignedContributionAndProof<T::EthSpec> {
+        &self.signed_aggregate
+    }
+}
+
+impl VerifiedSyncSignature {
+    /// Returns `Ok(Self)` if the `sync_signature` is valid to be (re)published on the gossip
+    /// network.
+    ///
+    /// `subnet_id` is the subnet from which we received this sync signature. This function will
+    /// verify that it was received on the correct subnet.
+    pub fn verify<T: BeaconChainTypes>(
+        sync_signature: SyncCommitteeSignature,
+        subnet_id: Option<SyncSubnetId>,
+        chain: &BeaconChain<T>,
+    ) -> Result<Self, Error> {
+        // Ensure sync committee signature is for the current slot (within a
+        // MAXIMUM_GOSSIP_CLOCK_DISPARITY allowance).
+        //
+        // We do not queue future sync committee messages for later processing.
+        verify_propagation_slot_range(chain, &sync_signature)?;
+
+        // Sync signatures must be for a known block. If the block is unknown, we simply drop the
+        // sync committee message and do not delay consideration for later.
+        verify_head_block_is_known(chain, sync_signature.beacon_block_root)?;
+
+        let pubkey = chain
+            .validator_pubkey_bytes(sync_signature.validator_index as usize)?
+            .ok_or(Error::UnknownValidatorIndex(
+                sync_signature.validator_index as usize,
+            ))?;
+
+        let sync_committee = chain.head_sync_committee_next_slot()?;
+        let subnet_positions = sync_committee.subcommittee_positions_for_public_key(&pubkey)?;
+
+        if let Some(subnet_id) = subnet_id {
+            if !subnet_positions.contains_key(&subnet_id) {
+                return Err(Error::InvalidSubnetId {
+                    received: subnet_id,
+                    expected: subnet_positions.keys().cloned().collect::<Vec<_>>(),
+                });
+            }
+        };
+
+        /*
+         * The sync committee message is the first valid message received for the participating validator
+         * for the slot, sync_signature.slot.
+         */
+        let validator_index = sync_signature.validator_index;
+        for subnet_id in subnet_positions.keys() {
```

Comment:
I don't think we should loop the subnet IDs here, because this function should only verify the `SyncCommitteeSignature` for the `subnet` provided as an argument. If a validator is in multiple subcommittees we don't want to mark it as seen for all subnets as soon as we see its signature on one subnet.

To simplify things, I'm pretty sure we can make the `subnet` argument `SyncSubnetId` rather than `Option<SyncSubnetId>`. We only verify sync committee signatures in two places:

* When receiving them via gossip, in which case the subnet is known
* When attempting to publish them from the VC via the API. In this case I was doing a dodgy thing where I called `verify` with a `None` subnet, but I think duplicate membership is rare enough that we should just absorb the cost of verifying multiple times with all the relevant subnets (which will almost always be 1 on mainnet).

Reply:
Right my bad. This would've been pretty much the same as not keying `observed_contributors` on `SyncSubnetId`. Updated :+1: